### ISSUE TITLE
Native LoRA fine-tune worker crashes with SIGABRT, independent of dataset size / batch size (reproducible 5/5 attempts)

### ISSUE THREAD

**Issue by alexbelij** [tetherto/qvac#3163]
**Environment:** Kaggle notebook, Tesla P100-PCIE-16GB GPU, `@qvac/sdk` (latest as of 2026-07-08), model: `Qwen3-1.7B-Q4_0.gguf`.

**Summary:** Calling the native LoRA fine-tune worker (`finetune()`) reliably crashes the native RPC worker process with `SIGABRT` before writing `adapter_meta.json`, even with the smallest possible workload.

**Steps to reproduce (from our project, MISTER — a Tether Developers Cup hackathon entry):**
1. `loadModel({ modelSrc: 'Qwen3-1.7B-Q4_0.gguf', modelType: 'llamacpp-completion' })`
2. Run BEFORE eval (`completion()`) — completes successfully every time.
3. Call `finetune({ model: modelId, ... })` with SFT pairs.

**Observed:**
- v2/v3 runs (93-pair dataset, default batch size): fine-tune starts for real, writes real checkpoints (`checkpoint_step_00000001`, `_00000002` — genuine `model.gguf` + `optimizer.gguf`) with genuine decreasing loss (`8.9185` -> `9.0051`), then crashes on the next step with:
  `WORKER_CRASHED: Bare worker exited mid-request (code=null, signal=SIGABRT)`
- v4/v5 runs (reduced to 20-pair dataset, `batchSize: 1`, `microBatchSize: 1` — smallest possible workload): crashes on the **very first** fine-tune call, before any checkpoint is written, **100% reproducible across 5 consecutive attempts** with fresh model reloads (`qvac.loadLLM`) before each retry (ruling out stale `modelId`/`MODEL_NOT_FOUND` as the cause).

**Conclusion:** the crash is independent of dataset size, batch size, and profile — this points to an intermittent/resource-related instability in the native LoRA fine-tune RPC worker itself (possibly a native addon/memory issue), not something fixable from the calling application. We built a retry/reload wrapper around `finetuneRun()` which correctly reloads the model and retries with backoff, but all 5 attempts still hit the same crash.

Happy to share full run logs / notebook if useful for reproduction.

**Comment by victorchimakanu** [comment]
The crash pattern — SIGABRT in the native RPC worker, reproducible 5/5 regardless of dataset size or batch size — points to a native addon stability issue rather than anything in your calling code. The fact that it hits even with a single pair and `batchSize: 1` rules out an OOM or dataset-shape problem.

Please share the full run logs. The stack trace from the SIGABRT will tell the team which native function is failing. If the Kaggle notebook captures stderr from the worker process, include that too.

A few things that would help narrow it down:
- Exact `@qvac/sdk` version (`npm list @qvac/sdk`)
- Whether the crash is consistent across fresh kernel restarts (rules out any GPU state leak between runs)
- Whether the pre-finetune `completion()` call you mention uses the same modelId or a different one

Flagging this to the team now.

**Comment by alexbelij** [comment]
Thanks for looking into this, @victorchimakanu. Here's the information you asked for.

### Environment

| Key | Value |
|-----|-------|
| `@qvac/sdk` | `0.14.1` (resolved via `package-lock.json`) |
| `@qvac/llm-llamacpp` | `0.30.1` (transitive dep) |
| Node.js | v20.20.2 |
| npm | 10.8.2 |
| GPU | Tesla P100-PCIE-16GB (16 GB) |
| NVIDIA Driver | 580.159.04 |
| CUDA | 13.0 |
| Platform | Kaggle Notebooks (Docker, `gcr.io/kaggle-private-byod/python`) |
| Model | `registry://hf/unsloth/Qwen3-1.7B-GGUF/.../Qwen3-1.7B-Q4_0.gguf` |
| LoRA targets | `attn_q, attn_k, attn_v, attn_o, ffn_gate, ffn_up, ffn_down` |

### Answers to your questions

**1. Full logs**

Below is the condensed v5 run log (the most recent). Full raw output is available on request.

<details>
<summary>v5 run log (click to expand)</summary>

```
[  500.5s] [INFO] [qvac] Health check passed
[  500.5s] [INFO] [finetune] Profile loaded {"profile":"gate_micro","epochs":1}
[  500.5s] [INFO] [finetune] SFT pairs loaded {"count":93}
[  500.5s] [INFO] [finetune] Truncating SFT pairs to profile maxSFTPairs {"before":93,"after":20}
[  500.5s] [INFO] [finetune] Causal corpus loaded {"count":20}
[  508.5s] [INFO] [qvac] LLM loaded {"modelId":"62ad5ed038880655"}
[  508.5s] [INFO] [finetune] Model info {"type":"llamacpp-completion","handlers":["completionStream","finetune","translate"],"isDelegated":false}
[  508.6s] [INFO] [finetune] Starting fine-tuning {"sftPairs":20,"causalDocs":20,"epochs":1}
[  508.6s] [INFO] [qvac] Wrote SFT training data {"path":"data/finetune_input/train.jsonl","pairs":20}
[  508.6s] [INFO] [qvac] Starting fine-tune {"modelId":"62ad5ed038880655","trainDir":"data/finetune_input","outputDir":"adapters","epochs":1,"sftPairs":20}

--- Attempt 1 ---
[  510.8s] [ERROR] finetuneRun attempt failed {"attempt":1,"crashed":true,"modelLost":false,"error":"Bare worker exited mid-request (code=null, signal=SIGABRT) — in-flight calls were aborted"}
[  510.8s] [WARN] Retrying after native worker crash {"nextAttempt":2,"resume":false,"backoffMs":3000}

--- Attempt 2 (model reloaded) ---
[  523.5s] [INFO] Reloaded model after crash {"modelId":"62ad5ed038880655"}
[  525.7s] [ERROR] finetuneRun attempt failed {"attempt":2,"crashed":true,"modelLost":false,...}

--- Attempt 3 (model reloaded) ---
[  541.5s] [INFO] Reloaded model after crash {"modelId":"62ad5ed038880655"}
[  543.8s] [ERROR] finetuneRun attempt failed {"attempt":3,"crashed":true,"modelLost":false,...}

--- Attempt 4 (model reloaded) ---
[  562.7s] [INFO] Reloaded model after crash {"modelId":"62ad5ed038880655"}
[  564.9s] [ERROR] finetuneRun attempt failed {"attempt":4,"crashed":true,"modelLost":false,...}

--- Attempt 5 (model reloaded) ---
[  587.0s] [INFO] Reloaded model after crash {"modelId":"62ad5ed038880655"}
[  589.4s] [ERROR] finetuneRun attempt failed {"attempt":5,"crashed":true,"modelLost":false,...}

--- Final error ---
[  589.4s] WORKER_CRASHED: Bare worker exited mid-request (code=null, signal=SIGABRT) — in-flight calls were aborted
    at handlePostHandshakeExit (node_modules/@qvac/sdk/dist/client/rpc/node-rpc-client.js:238:28)
    at ChildProcess.<anonymous> (node_modules/@qvac/sdk/dist/client/rpc/node-rpc-client.js:352:25)
    at ChildProcess.emit (node:events:524:28)
    at ChildProcess._handle.onexit (node:internal/child_process:293:12)
```

</details>

The crash is immediate — the worker exits with SIGABRT ~2 seconds after `finetuneRun()` is called. Every single time.

**2. Exact SDK version**

`@qvac/sdk@0.14.1` with `@qvac/llm-llamacpp@0.30.1` (both from `package-lock.json`).

**3. Fresh kernel restarts**

Yes. We ran 5 separate kernel versions (v1–v5, on different dates over two weeks). The v5 run shown above also does explicit fresh `loadLLM()` calls between every retry attempt. The crash is 100% reproducible.

**4. Same `modelId` for `completion()` and `finetune()`?**

Yes, same flow:
```js
const modelId = await qvac.loadLLM(MODEL, { quantization: 'Q4_0', ctxSize: 2048 });
// completion() works fine — we ran 5 eval prompts successfully before finetune
await qvac.finetuneRun(modelId, params, onProgress);  // ← always crashes here
```

`completion()` works perfectly — the model generates real responses. Only `finetuneRun()` triggers the SIGABRT.

### Additional observations

- In earlier runs (v2, v3) with larger batch sizes, the worker occasionally survived 1–2 training steps (producing real checkpoints with decreasing loss values) before crashing. In v4/v5 with minimal settings (`batchSize: 1`, `microBatchSize: 1`, 20 SFT pairs, 1 epoch), it crashes before the first step completes.

- We initially had two bugs in our own code (CLI parser and batch size wiring) that we've since fixed. The SIGABRT persists regardless of the actual workload size — it happens even at the absolute minimum configuration.

### Possibly relevant: recent commits on `main`

I noticed these were merged today on the `tetherto/qvac` main branch:

- **#3148** `fix: map bare-rpc channel-closed errors to worker error classes` — directly describes `bare-rpc` rejecting with `CHANNEL_CLOSED`, which could race with worker crash detection
- **#3159** `fix: synchronize async llama decode completion` — race condition fixes in GPU decode
- `llm-llamacpp` bumped from 0.30.1 → 0.36.3

These aren't released yet on npm (latest is still 0.14.1). If there's a way to test against a pre-release build from `main`, I'm happy to re-run the fine-tune and report back.

---

We're using QVAC for our Tether Developers Cup hackathon project (on-device LoRA fine-tuning for football coaching analytics). The fine-tune pipeline is a core part of the submission, so any guidance on timeline or workaround would be very appreciated.

**Comment by victorchimakanu** [comment]
The logs are clear and the investigation is solid. The crash at ~2s into every `finetuneRun()` call — before the first training step — is a native addon fault in the fine-tune worker, not anything in your calling code.

Your read on the recent main commits is correct. Both #3148 (bare-rpc channel-closed errors mapped to worker error classes) and #3159 (async llama decode synchronization) address crash-path issues in the same native layer. The `@qvac/llm-llamacpp` package moved significantly on main since the 0.14.1 SDK release, and your installed version (0.30.1, pinned by the SDK) predates those fixes.

The fixes are merged on main but not yet in the published npm release. I am flagging this to the team with your hackathon deadline context — the Tether Developers Cup fine date is July 12. Testing against a pre-release main build from a Kaggle notebook is not straightforward, so the best path right now is to watch for the next SDK release on npm and re-run the fine-tune at that point.

Will update this thread when a new release is available.

**Comment by victorchimakanu** [comment]
@qvac/sdk 0.15.0 is on npm as of this morning. It carries both fixes from the earlier investigation. The worker error mapping from #3148 ships inside the SDK package itself, and the async decode synchronization from #3159 is in @qvac/llm-llamacpp 0.36.3, which 0.15.0 pulls in. Your lockfile resolved llm-llamacpp at 0.30.1, so the old native worker is what crashed.

To retest, delete node_modules and package-lock.json, set @qvac/sdk to 0.15.0, run a fresh npm install, and rerun your smallest finetuneRun() case, the single pair with batchSize 1. If the SIGABRT still reproduces on 0.15.0, post the new stack trace here and the team will take it further.

**Comment by victorchimakanu** [comment]
Both fixes are in @qvac/sdk 0.15.0, which is live on npm. To pick them up, clear your lockfile and reinstall:

```
rm -rf node_modules package-lock.json
npm install
```

Your lockfile had pinned `@qvac/llm-llamacpp` at 0.30.1. The new install will resolve 0.36.3, which carries the async decode synchronization fix from #3159. If the SIGABRT persists after a clean install on 0.15.0, reopen with the updated logs.
